internal/leasecontroller/lease_controller.go

```diff
@@ -0,0 +1,335 @@
+	getNodeFunc func(context.Context) (*corev1.Node, error)) LeaseController {
+
+	if leaseRenewalInterval == 0 {
+		panic("Lease renewal interval is 0")
+	}
+
+	lc := &v1Betav1LeaseController{
+		client:               client,
+		leaseRenewalInterval: leaseRenewalInterval,
+		pingStatusFunc:       pingStatusFunc,
+		getNodeFunc:          getNodeFunc,
+	}
+	if lease == nil {
+		lc.lease = &coord.Lease{}
+	} else {
+		lc.lease = lease.DeepCopy()
+	}
+
+	return lc
+}
+
+type v1Betav1LeaseController struct {
+	client               v1beta1.LeaseInterface
+	leaseRenewalInterval time.Duration
+	pingStatusFunc       func(context.Context) (*nodepingcontroller.PingResult, error)
+	getNodeFunc          func(context.Context) (*corev1.Node, error)
+	lease                *coord.Lease
+
+	wsStatusLock sync.Mutex
+	workStatus   Status
+}
+
+func (lc *v1Betav1LeaseController) LeaseControllerStatus(ctx context.Context) Status {
+	lc.wsStatusLock.Lock()
+	defer lc.wsStatusLock.Unlock()
+	return lc.workStatus
+}
+
+func (lc *v1Betav1LeaseController) Run(ctx context.Context) {
+	ctx = log.WithLogger(ctx, log.G(ctx).WithField("leaseRenewalInterval", lc.leaseRenewalInterval))
+	for {
+		sleepTime := lc.leaseRenewalInterval
+
+		err := lc.poll(ctx)
+		if err != nil {
+			if pkgerrors.Is(err, context.Canceled) {
+				return
+			}
+
+			lc.wsStatusLock.Lock()
+			lastWorkStatus := lc.workStatus
+			if lastWorkStatus == Pending && !pkgerrors.Is(err, &nodeNodeReadyError{}) {
+				lc.workStatus = FailedToEstablish
+			}
+			lc.wsStatusLock.Unlock()
+
+			// This should only happen if the lease had never been established before and if we think the API server
+			// does not support leases
+			if pkgerrors.Is(err, &shutdownError{}) {
+				if lastWorkStatus != Pending {
+					panic(fmt.Sprintf("Lease controller shutting down after lease established: %v", err))
+				}
+				log.G(ctx).WithError(err).Error("Lease controller shutting down")
+				return
+			}
+
+			if seconds, delay := errors.SuggestsClientDelay(err); delay {
+				sleepTime = time.Second * time.Duration(seconds)
+			}
+			log.G(ctx).WithError(err).WithField("sleepTime", sleepTime).Warn("Failed to update lease. Retrying")
+		} else {
+			lc.wsStatusLock.Lock()
+			lc.workStatus = Established
+			lc.wsStatusLock.Unlock()
+		}
+		timer := time.NewTimer(sleepTime)
+		select {
+		case <-timer.C:
+		case <-ctx.Done():
+			timer.Stop()
+			return
+		}
+	}
+}
+
+func (lc *v1Betav1LeaseController) poll(ctx context.Context) (retErr error) {
+	ctx, span := trace.StartSpan(ctx, "v1Betav1LeaseController.poll")
+	defer span.End()
+
+	defer func() {
+		span.SetStatus(retErr)
+	}()
+
+	pr, err := lc.pingStatusFunc(ctx)
+	if err != nil {
+		return fmt.Errorf("Received error when attempting to ascertain node status: %w", err)
+	}
+
+	if pr.Error != nil {
+		return newNodeNodeReady(pr)
+	}
+
+	lc.lease.Spec.RenewTime = &metav1.MicroTime{Time: time.Now()}
+	// This is 25 due to historical reasons. It was supposed to be * 5, but...reasons
```

Comment:
What reasons?

Reply:
someone multiplied times 5 on input..and then on 5 again. In my limited testing in prod, suddenly changing this value doesn't seem to make other components happy.